#### Life cycle of Machine learning Project

- Understanding the Problem Statement
- Data Collection
- Data Checks to perform
- Exploratory data analysis
- Data Pre-Processing
- Model Training
- Choose best model

### 1) Problem statement
- This project understands how the student's performance (test scores) is affected by other variables such as Gender, Ethnicity, Parental level of education, Lunch and Test preparation course.


### 2) Data Collection
- Dataset Source - https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?datasetId=74977
- The data consists of 8 column and 1000 rows.

### 2.1 Import Data and Required Packages
####  Importing Pandas, Numpy, Matplotlib, Seaborn and Warings Library.


In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

### Import the CSV dataset as pandas DF


In [3]:
df = pd.read_csv("data/stud.csv")

In [4]:
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [5]:
df.shape

(1000, 8)

### 2.2 Dataset information
- gender : sex of students  -> (Male/female)
- race/ethnicity : ethnicity of students -> (Group A, B,C, D,E)
- parental level of education : parents' final education ->(bachelor's degree,some college,master's degree,associate's degree,high school)
- lunch : having lunch before test (standard or free/reduced) 
- test preparation course : complete or not complete before test
- math score
- reading score
- writing score

### 3. Data checks to perform
Check missing values

Check Duplicates

Check data types


Check the number of unique values of each columns

Check the statistics of data set

Check the various categories present in the different cat. columns

In [12]:
df.isna().sum()

gender                         0
race_ethnicity                 0
parental_level_of_education    0
lunch                          0
test_preparation_course        0
math_score                     0
reading_score                  0
writing_score                  0
dtype: int64

In [15]:
df.duplicated().sum()

0

In [17]:
df.dtypes

gender                         object
race_ethnicity                 object
parental_level_of_education    object
lunch                          object
test_preparation_course        object
math_score                      int64
reading_score                   int64
writing_score                   int64
dtype: object

In [18]:
df.nunique()

gender                          2
race_ethnicity                  5
parental_level_of_education     6
lunch                           2
test_preparation_course         2
math_score                     81
reading_score                  72
writing_score                  77
dtype: int64

In [19]:
df.describe()

math_score  reading_score  writing_score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

#### Insight
- From above description of numerical data, all means are very close to each other - between 66 and 68.05;
- All standard deviations are also close - between 14.6 and 15.19;
- While there is a minimum score  0 for math, for writing minimum is much higher = 10 and for reading myet higher = 17

In [25]:
e = df["gender"].nunique()
print(f"Categories in Gender column : {e}", end=" ")
print(df["gender"].unique())

e = df["lunch"].nunique()
print(f"Categories in lunch column : {e}", end=" ")
print(df["lunch"].unique())

e = df["parental_level_of_education"].nunique()
print(f"Categories in parental_level_of_education column : {e}", end=" ")
print(df["parental_level_of_education"].unique())

e = df["race_ethnicity"].nunique()
print(f"Categories in race_ethnicity column : {e}", end=" ")
print(df["race_ethnicity"].unique())

e = df["test_preparation_course"].nunique()
print(f"Categories in test_preparation_course column : {e}", end=" ")
print(df["test_preparation_course"].unique())


Categories in Gender column : 2 ['female' 'male']
Categories in lunch column : 2 ['standard' 'free/reduced']
Categories in parental_level_of_education column : 6 ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
Categories in race_ethnicity column : 5 ['group B' 'group C' 'group A' 'group D' 'group E']
Categories in test_preparation_course column : 2 ['none' 'completed']


In [28]:
#Define Numerical and Categorical columns

num_col=[feature for feature in df.columns if df[feature].dtype !='O']
cat_col=[feature for feature in df.columns if df[feature].dtype == 'O']

print("We have {} numerical features : {}".format(len(num_col),num_col))
print("We have {} Categorical features : {}".format(len(cat_col),cat_col))

We have 3 numerical features : ['math_score', 'reading_score', 'writing_score']
We have 5 Categorical features : ['gender', 'race_ethnicity', 'parental_level_of_education', 'lunch', 'test_preparation_course']


In [29]:
df["Total_score"] = df["math_score"] + df["reading_score"] + df["writing_score"]
df["AVG"] = df["Total_score"]/3
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   Total_score        AVG  
0          218  72.666667  
1          247  82.333333  
2          278  92.666667  
3          148  49.333333  
4          229  76.333333

In [31]:
reading_full = df[df['reading_score'] == 100]['AVG'].count()
writing_full = df[df['writing_score'] == 100]['AVG'].count()
math_full = df[df['math_score'] == 100]['AVG'].count()

print(f'Number of students with full marks in Maths: {math_full}')
print(f'Number of students with full marks in Writing: {writing_full}')
print(f'Number of students with full marks in Reading: {reading_full}')

Number of students with full marks in Maths: 7
Number of students with full marks in Writing: 14
Number of students with full marks in Reading: 17


In [ ]:
reading_full = df[df['reading_score'] == 100]['AVG'].count()
writing_full = df[df['writing_score'] == 100]['AVG'].count()
math_full = df[df['math_score'] == 100]['AVG'].count()

print(f'Number of students with full marks in Maths: {math_full}')
print(f'Number of students with full marks in Writing: {writing_full}')
print(f'Number of students with full marks in Reading: {reading_full}')